### ISSUE TITLE
Invite Flow: empty avatar for system entry (Apple) and unnamed entries

### ISSUE THREAD

**Issue by riyapenn** [WhisperSystems/Signal-iOS#1517]
### Bug description
Empty avatars in invite flow for ALL scenarios below
1. the contact is an apple system entry
2. in message view, scroll to '#' see empty info for contact entry who has just an email address nothing else (i.e. no name)
3. in mail view, scroll to '#' see empty info for contact entry who has just a phone number nothing else (i.e. no name)

### Steps to reproduce
- Have contact entry for Apple, Inc as set by Apple
- Have contact entry of solely a phone number, no other information in iOS contact list; Then View Invite Flow  > Message 
- Have contact entry of solely an email address, no other information in iOS contact list; Then View Invite Flow  > Mail 

**Actual result:** Empty avatar displayed
**Expected result:** Suggestion, but not definite: '#' avatar displayed (https://github.com/WhisperSystems/Signal-iOS/issues/1512 should relate to what is displayed here)

### Screenshots

Sample with entry of solely email and system 
![inviteflownoavatar](https://cloud.githubusercontent.com/assets/4898366/20819641/9500e7fc-b7ea-11e6-8c16-a339db0c08eb.png)

### Device info

**Device**: iPhone 5C and 6S
**iOS version**: 10.1.1
**Signal version:** 2.6.6

**Comment by michaelkirk** [comment]
Not strictly a dupe, but seems like @roostr solution in https://github.com/WhisperSystems/Signal-iOS/issues/1512 will solve this.

**Comment by roostr** [comment]
Updated status after changes in #1512:

Avatars:
 * ✅  Contacts that have a name but no phone number are listed with either a "#" avatar or the contact image, depending on availability
 * ✅  Contacts that are just a phone number have a "#" avatar
 * ❌  Contacts that have a "business name" (like Apple Inc) have an avatar that's a blank colored circle, no initials nor "#" inside

Ordering:
 * ✅  Contacts that have a name but no phone number are listed in the correct section of the list (eg. Brad Williams is listed in the W section)
 * ✅  Contacts that are just a phone number are listed in the "#" section
 * ❌  Contacts that have a "business name" (like Apple Inc) are listed in the "#" section

Expected results (I think):
 * Contacts that have a "business name" should be listed under the letter of the business name (eg. Apple should be in the A section)
 * Contacts that have a "business name" (like "Apple Inc") should have an "initialization" of their business name in the circle (eg "Apple Inc" would be "AI")

**Comment by roostr** [comment]
Three fixes were needed, here's the proposed solution:

1. Organizations being sorted to the bottom under the "#" section is fixed in 144a89e.
2. The first step of the fix for contacts having "#" as an avatar instead of the initials is in 74ebcfe and also in [SignalServiceKit-1f94d29](https://github.com/roostr/SignalServiceKit/commit/1f94d29616dfa6fb3727dfba8e678484a525770b). It was happening because my recent avatar code was asking ContactsManager to look up a contact, then report whether its name was generated from a phone number. It was returning NO for contacts that weren't found in CM, which happened for some contacts that were built straight from CNContact but not in CM. These two commits instead evaluate whether the name was built from the phone number on the Contact object itself.
3. The previous solution wasn't quite enough, as I found that some business contacts from CNContact still didn't have the proper initials in their avatars. To fix that, [SignalServiceKit-8167706](https://github.com/roostr/SignalServiceKit/commit/816770681380d7510e8073f9e52dc7e68974d30b). It just duplicates some fixups that are already being done for ABAddressBook-sourced contacts via OWSContactsManager. I realize this code is probably ripe for a refactor, but decided to leave TODOs for future cleanup rather than fix it now, as I don't have the time to verify I don't introduce regressions with a RF, but would love to see this invite flow feature move closer to ready for release.

**Comment by michaelkirk** [comment]
FYI I did make some changes to the avatar/contact logic here: #1523. 

>  I realize this code is probably ripe for a refactor

I think the same could be said for all of Contact/OWSContactManager. In particular, I think including AddressBook.framework and Contact.framework logic in SSK was a mistake.

I started to replace calls to SSK:Contact.m with S-iOS:ContactAdapter here: https://github.com/WhisperSystems/Signal-iOS/blob/mkirk/wip-contact-adapters/Signal/src/contact/ContactAdapter.swift

But it started to look like it was going to be a big thing, and I cowered away. Maybe next time I'm stuck on a plane?
